### PR TITLE
addrman: Fix new table bucketing during unserialization

### PR DESCRIPTION
**Pull request by jnewbery**
This fixes three issues in addrman unserialization.

1. An addrman entry can appear in up to 8 new table buckets. We store this entry->bucket indexing during shutdown so that on restart we can restore the entries to their correct buckets. Commit ec45646de9e62b3d42c85716bfeb06d8f2b507dc broke the deserialization code so that each entry could only be put in up to one new bucket.

2. Unserialization may result in an entry appearing in a 9th bucket. If the entry already appears in 8 buckets don't try to place it in another bucket.

3. We unnecessarily rebucket when reading a peers.dat with file version 1. Don't do that.

### CODE REVIEW

**Comment by jnewbery** [comment]
The code currently rebuckets entries (losing the additional bucket positions) if the asmap checksum changes. I'm not sure if that's necessary and don't see any downside to keeping the same bucketing if restarting with a new asmap. If others agree, I can update this PR to not do that rebucketing.

**Comment by jnewbery** [comment]
We should also add a regression test for this. I think a good test would be to add a round trip serialize/unserialize to the addrman fuzz test, and compare the addrman before and after that round trip. There's some internal state that isn't reconstructed from serialize/unserialize (eg the vRandom ordering), but a comparator function that checked number of entries and bucket->entry pairs should work.

If anyone wants to add that fuzz test or other tests, I'm happy to add them to this PR.

File: src/addrman.h [hunk 1]
```diff
@@ -500,47 +501,59 @@ friend class CAddrManTest;
...
-            supplied_asmap_version = SerializeHash(m_asmap);
+            supplied_asmap_checksum = SerializeHash(m_asmap);
         }
-        uint256 serialized_asmap_version;
+        uint256 serialized_asmap_checksum;
         if (format >= Format::V2_ASMAP) {
-            s >> serialized_asmap_version;
+            s >> serialized_asmap_checksum;
         }
+        const bool restore_bucketing{nUBuckets == ADDRMAN_NEW_BUCKET_COUNT &&
```
Comment (MarcoFalke) [inline src/addrman.h:530]:
```suggestion
        const bool restore_bucketing{nUBuckets == ADDRMAN_NEW_BUCKET_COUNT &&
```

Comment (jnewbery) [reply src/addrman.h:530]: fixed

**Comment by naumenkogs** [comment]
Concept ACK.

>The code currently rebuckets entries (losing the additional bucket positions) if the asmap checksum changes. I'm not sure if that's necessary and don't see any downside to keeping the same bucketing if restarting with a new asmap. If others agree, I can update this PR to not do that rebucketing.

That would mean that asmap is applied to only newly learned addresses, which is a downside (assuming asmap is an improvement). On the other hand, losing the existing 8 buckets is unfortunate.
I'm fine either way.

We could do the 8 existing + 1 asmap bucket if we really want... but that also has some consequences of potentially overriding existing buckets etc.

Also, we should keep in mind that non-rebucketing on a new asmap may allow existing records to appear in 16 buckets now instead of 8 (probably *all* existing records on first asmap appearance). Which is... also not particularly terrible but should be considered?

**Comment by jnewbery** [comment]
> That would mean that asmap is applied to only newly learned addresses, which is a downside (assuming asmap is an improvement)

Currently the bucket positions for existing new addresses are lost entirely when changing asmap, so I don't understand how this is a downside. The choice is between throwing away those bucket positions or not.

> Also, we should keep in mind that non-rebucketing on a new asmap may allow existing records to appear in 16 buckets now instead of 8 (probably all existing records on first asmap appearance). Which is... also not particularly terrible but should be considered?

No. We won't add an entry to more than 8 buckets: https://github.com/bitcoin/bitcoin/blob/257cf05f9b841ba30202f23a94bcdb1743feded2/src/addrman.cpp#L294-L296.

**Comment by naumenkogs** [comment]
> Currently the bucket positions for existing new addresses are lost entirely when changing asmap, so I don't understand how this is a downside. The choice is between throwing away those bucket positions or not.

I agree with you in the next sentence: "On the other hand, losing the existing 8 buckets is unfortunate.". Still, ideally, we would apply a new asmapw to existing addrs without losing the data. Unfortunately, since we don't store addr source, we have to make this choice. And I agree that of these 2 choices, the one you're suggesting is probably better.

>No. We won't add an entry to more than 8 buckets: 

Good to know! That was just my wrong assumption without looking deep into code :)

File: src/addrman.h [hunk 2]
```diff
@@ -348,8 +349,8 @@ friend class CAddrManTest;
      * Notice that vvTried, mapAddr and vVector are never encoded explicitly;
      * they are instead reconstructed from the other information.
      *
-     * vvNew is serialized, but only used if ADDRMAN_UNKNOWN_BUCKET_COUNT didn't change,
-     * otherwise it is reconstructed as well.
+     * vvNew is serialized, but only used if ADDRMAN_UNKNOWN_BUCKET_COUNT and the asmap checksum
```
Comment (naumenkogs) [inline src/addrman.h:352]: This comment is weird in general. I don't think we have anything like `ADDRMAN_UNKNOWN_BUCKET_COUNT` anymore?

Comment (naumenkogs) [reply src/addrman.h:352]: Should be talking about ADDRMAN_NEW_BUCKET_COUNT, ADDRMAN_NEW_BUCKETS_PER_ADDRESS.

File: src/addrman.h [hunk 3]
```diff
@@ -525,19 +527,26 @@ friend class CAddrManTest;
...
 
         for (auto bucket_entry : bucket_entries) {
             int bucket{bucket_entry.first};
             const int entry_index{bucket_entry.second};
             CAddrInfo& info = mapInfo[entry_index];
+
+            // The entry shouldn't appear in more than
+            // ADDRMAN_NEW_BUCKETS_PER_ADDRESS. If it has already, just skip
+            // this bucket_entry.
+            if (info.nRefCount >= ADDRMAN_NEW_BUCKETS_PER_ADDRESS) continue;
```
Comment (naumenkogs) [inline src/addrman.h:541]: So this happens if a new version of Bitcoin Core reduced `ADDRMAN_NEW_BUCKETS_PER_ADDRESS`, and in this case we completely forget the extra addrs.

Why do you think it's more reasonable? Assuming addrman is less-than-half-full, there's a good chance this odd addr can find a new place without evicting anything.

File: src/addrman.h [hunk 4]
```diff
@@ -517,6 +517,8 @@ friend class CAddrManTest;
             }
         }
 
+        // Attempt to restore the entry's new buckets if the bucket count and asmap
```
Comment (naumenkogs) [inline src/addrman.h:520]: I find it hard to parse "restore the entry's new buckets".
1. Not only buckets, but also positions.
2. What exactly is "restore" is also a bit unclear, I'm thinking if there's a better verb. Perhaps just "use the serialized [...]"
3.

File: src/addrman.h (hunk 2 above)
Comment (jnewbery) [reply src/addrman.h:352]: Thanks, changed to ADDRMAN_NEW_BUCKET_COUNT.

File: src/addrman.h (hunk 3 above)
Comment (jnewbery) [reply src/addrman.h:541]: Because `ADDRMAN_NEW_BUCKETS_PER_ADDRESS` should be an invariant in addrman. No entry should appear in no more than `ADDRMAN_NEW_BUCKETS_PER_ADDRESS` buckets.

**Comment by jnewbery** [comment]
Thanks for the review @naumenkogs. I've addressed all your comments.

**Comment by laanwj** [comment]
Concept ACK, and the code change looks overall good to me. 

> https://github.com/bitcoin/bitcoin/pull/20557#issuecomment-737890278

Agree that a test would be useful, to make it easier to see that this fixes the mentioned problems and prevent future regression.

**Comment by naumenkogs** [comment]
utACK dbdf63c6d82ae8d1ca73f11d1b8844614f0bf8e3

File: src/addrman.h [hunk 5]
```diff
@@ -504,14 +504,14 @@ friend class CAddrManTest;
...
-            int nSize = 0;
-            s >> nSize;
-            for (int n = 0; n < nSize; n++) {
-                int nIndex = 0;
-                s >> nIndex;
-                if (nIndex >= 0 && nIndex < nNew) {
-                    bucket_entries.emplace_back(bucket, nIndex);
+        for (int bucket = 0; bucket < nUBuckets; ++bucket) {
+            int bucket_size{0};
+            s >> bucket_size;
```
Comment (glozow) [inline src/addrman.h:510]: Just to clarify... this is the number of entries in this bucket, right? Not the capacity of the bucket, which is `ADDRMAN_BUCKET_SIZE`?

File: src/addrman.h [hunk 6]
```diff
@@ -525,19 +527,26 @@ friend class CAddrManTest;
...
             const int entry_index{bucket_entry.second};
             CAddrInfo& info = mapInfo[entry_index];
+
+            // The entry shouldn't appear in more than
+            // ADDRMAN_NEW_BUCKETS_PER_ADDRESS. If it has already, just skip
+            // this bucket_entry.
+            if (info.nRefCount >= ADDRMAN_NEW_BUCKETS_PER_ADDRESS) continue;
+
             int nUBucketPos = info.GetBucketPosition(nKey, true, bucket);
-            if (format >= Format::V2_ASMAP && nUBuckets == ADDRMAN_NEW_BUCKET_COUNT && vvNew[bucket][nUBucketPos] == -1 &&
```
Comment (glozow) [inline src/addrman.h:530]: So you're removing the `format >= Format::V2_ASMAP` condition here because otherwise it'd be impossible for `serialized_asmap_checksum == supplied_asmap_checksum` yes?

**Review by glozow — COMMENTED** [review]
code review ACK https://github.com/bitcoin/bitcoin/commit/dbdf63c6d82ae8d1ca73f11d1b8844614f0bf8e3

Can see that it fixes the bug from https://github.com/bitcoin/bitcoin/commit/ec45646de9e62b3d42c85716bfeb06d8f2b507dc, previously looped over each entry, retrieving the first bucket we recorded in `entryToBucket`, and now creates a list of `<bucket, index>` pairs and loops over all of those instead. Some clarifying questions because I'm super unfamiliar with this code and stared at it for a long time.

File: src/addrman.h (hunk 5 above)
Comment (jnewbery) [reply src/addrman.h:510]: Exactly (see the comment above about serialization format: https://github.com/jnewbery/bitcoin/blob/dbdf63c6d82ae8d1ca73f11d1b8844614f0bf8e3/src/addrman.h#L341)

File: src/addrman.h (hunk 6 above)
Comment (jnewbery) [reply src/addrman.h:530]: Correct. If format == V1_DETERMINISTIC, then serialized_asmap_checksum must be 0.

File: src/addrman.h (hunk 5 above)
Comment (glozow) [reply src/addrman.h:510]: I just thought both being called "size" is a bit confusing so `num_entries` would have been my choice for naming, but it's the nittiest of nits, non-blocking

Comment (jnewbery) [reply src/addrman.h]: I agree that this is confusing. I've renamed to `num_entries`.

**Comment by jnewbery** [comment]
Improved the naming of one of the variables as suggested in https://github.com/bitcoin/bitcoin/pull/20557#discussion_r556985482 :
 https://github.com/bitcoin/bitcoin/compare/dbdf63c6d82ae8d1ca73f11d1b8844614f0bf8e3..36f3d9d8a48541f0b44ea4cff52a770cab43f2f0.

**Comment by jnewbery** [comment]
Rebased on master

**Comment by naumenkogs** [comment]
ACK ac3547eddd8a7d67b4103508f30d5d02a9c1f148

File: src/addrman.h (hunk 3 above)
Comment (vasild) [reply src/addrman.h:542]: @naumenkogs if a new Bitcoin Core reduces `ADDRMAN_NEW_BUCKETS_PER_ADDRESS` from 8 to e.g. 6 and this condition becomes true in the new version, then the address is already in 6 buckets in memory. So we would not completely forget it, or do I miss something?

File: src/addrman.h [hunk 7]
```diff
@@ -500,47 +501,60 @@ friend class CAddrManTest;
...
                 // Bucketing has not changed, using existing bucket positions for the new table
-                vvNew[bucket][nUBucketPos] = n;
-                info.nRefCount++;
+                vvNew[bucket][bucket_position] = entry_index;
+                ++info.nRefCount;
             } else {
-                // In case the new table data cannot be used (format unknown, bucket count wrong or new asmap),
+                // In case the new table data cannot be used (bucket count wrong or new asmap),
                 // try to give them a reference based on their primary source address.
                 LogPrint(BCLog::ADDRMAN, "Bucketing method was updated, re-bucketing addrman entries from disk\n");
```
Comment (vasild) [inline src/addrman.h:552]: If `restore_bucketing` is `false`, then this will be printed for every address, possibly _many_ times. But it does not include the address itself. What about printing the address:

```suggestion
                LogPrint(BCLog::ADDRMAN, "Bucketing method was updated, re-bucketing addrman entry from disk for %s\n", info.ToString());
```

or print the generic message once:

```suggestion
                if (!printed) {
                    LogPrint(BCLog::ADDRMAN, "Bucketing method was updated, re-bucketing addrman entries from disk\n");
                    printed = true;
                }
```

**Comment by vasild** [comment]
> I think a good test would be to add a round trip serialize/unserialize to the addrman fuzz test, and compare the addrman before and after that round trip. There's some internal state that isn't reconstructed from serialize/unserialize (eg the vRandom ordering), but a comparator function that checked number of entries and bucket->entry pairs should work.

Created such test at: https://github.com/jnewbery/bitcoin/pull/18.

It does not compare bucket->entry though because the ids (the keys into `mapInfo`) could be different after ser/unser. Hmm...

File: src/addrman.h [hunk 8]
```diff
@@ -500,47 +501,60 @@ friend class CAddrManTest;
...
+            const int entry_index{bucket_entry.second};
+            CAddrInfo& info = mapInfo[entry_index];
+
+            // The entry shouldn't appear in more than
+            // ADDRMAN_NEW_BUCKETS_PER_ADDRESS. If it has already, just skip
+            // this bucket_entry.
+            if (info.nRefCount >= ADDRMAN_NEW_BUCKETS_PER_ADDRESS) continue;
+
+            int bucket_position = info.GetBucketPosition(nKey, true, bucket);
+            if (restore_bucketing && vvNew[bucket][bucket_position] == -1) {
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/addrman.h:545]: If both `restore_bucketing` is false and `vvNew[bucket][bucket_position] != -1` then it means that two addresses hash into the same position in the same new bucket, which is possible based on the source address. `restore_bucketing` would be false if the number of new buckets had changed or the as map had changed.

File: src/addrman.h (hunk 7 above)
Comment (jnewbery) [reply src/addrman.h:552]: Good catch! I'm just going to move the log to outside the loop to avoid it printing many times. Improved logging can be left for a future PR.

File: src/addrman.h [hunk 9]
```diff
@@ -348,8 +349,8 @@ friend class CAddrManTest;
      * Notice that vvTried, mapAddr and vVector are never encoded explicitly;
      * they are instead reconstructed from the other information.
      *
-     * vvNew is serialized, but only used if ADDRMAN_UNKNOWN_BUCKET_COUNT didn't change,
-     * otherwise it is reconstructed as well.
+     * vvNew is serialized, but only used if ADDRMAN_NEW_BUCKET_COUNT and the asmap checksum
+     * didn't change, otherwise it is reconstructed as well.
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/addrman.h:353]: Good suggestions. I've added a new commit that makes these changes.

**Comment by jnewbery** [comment]
Thanks for the review @vasild. I've resolved your review comments.

I haven't taken your fuzz test yet since it seems like there's a bit more work to make sure that it's getting reliable coverage.

File: src/addrman.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (vasild) [reply src/addrman.h]: Thanks! `s/vVector/vRandom/` in that commit - there is no member named `vVector`.
Also, consider adding `mapInfo` to the list of members that "are never encoded explicitly" - the ids (the keys of that map) are not serialized, only the values (and in different order than in the map).

File: src/addrman.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (vasild) [reply src/addrman.h:546]: Yes! So it follows that (correct me if I am wrong):

* If `restore_bucketing` is `true` then `vvNew[bucket][bucket_position] == -1` is also `true` (position is free, not occupied). So the `if` condition can be reduced to just `if (restore_bucketing) {`.

* If `restore_bucketing` is `false` then `vvNew[bucket][bucket_position] == -1` may be `true` or `false` depending on the hashing. In this case too, the `if` condition is equivalent to just `if (restore_bucketing) {`.

File: src/addrman.h (hunk 7 above)
Comment (vasild) [reply src/addrman.h]: `s/restore_bucketing/!restore_bucketing/` in 9a0760db4?

File: src/addrman.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/addrman.h:546]: I think `vvNew[bucket][bucket_position] == -1` is just a belt-and-suspenders. It shouldn't be hit (if so, then something has gone wrong with the position serializing/deserializing to place two addresses in the same bucket/position). I think you're right that we can just remove it, but can we leave that as a follow-up? We can't fix _all_ of addrman's problems in one PR :)

File: src/addrman.h (hunk 7 above)
Comment (jnewbery) [reply src/addrman.h]: :man_facepalming: You're right of course. I'll fix.

File: src/addrman.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/addrman.h]: Thanks. I applied the diff by hand and missed those. Now fixed!

**Comment by vasild** [comment]
ACK 4676a4fb5be0f6ef0b3f71c1f4361c20f7cb0e0b

* The addition of a fuzzing test (https://github.com/jnewbery/bitcoin/pull/18) is not a show stopper for this PR.
* To my understanding, the unneeded [`vvNew[bucket][bucket_position] == -1`](https://github.com/bitcoin/bitcoin/pull/20557/files#diff-164bd9e2e30f54d0a79eb7cc372309e2f2155edc6c3f051290ab078f03f6a771R546) can be removed. It was part of the code before this PR and is doing no harm (other than confusion). Not a show stopper either.

**Review by vasild — APPROVED** [review]

File: src/addrman.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (vasild) [reply src/addrman.h:546]: Alright, so either we are both correct that this is redundant and not needed, or we are both wrong!

If we are wrong, and it can happen that `restore_bucketing` is `true` and `vvNew[bucket][bucket_position] == -1` is `false` (occupied!) then moving the `LogPrint()` call was wrong because it would have been executed inside the `else` branch and now it will not be executed because now it is conditional only on `restore_bucketing`.

**Comment by glozow** [comment]
re-ACK https://github.com/bitcoin/bitcoin/commit/4676a4fb5be0f6ef0b3f71c1f4361c20f7cb0e0b, changes were a rename, comments, and removing repeat-logging.

**Comment by naumenkogs** [comment]
ACK 4676a4f

**Comment by jnewbery** [comment]
@MarcoFalke or @laanwj - this PR now has 3 ACKs. If one of you re-reviewed, then it might be ready for merge.

File: src/addrman.h [hunk 10]
```diff
@@ -335,23 +335,20 @@ friend class CAddrManTest;
...
      *
      * 2**30 is xorred with the number of buckets to make addrman deserializer v0 detect it
      * as incompatible. This is necessary because it did not check the version number on
      * deserialization.
      *
-     * Notice that vvTried, mapAddr and vVector are never encoded explicitly;
+     * vvNew, vvTried, mapInfo, mapAddr and vRandom are never encoded explicitly;
      * they are instead reconstructed from the other information.
      *
-     * vvNew is serialized, but only used if ADDRMAN_NEW_BUCKET_COUNT and the asmap checksum
```
Comment (dhruv) [inline src/addrman.h]: In the format above aren't lines 340-342 an encoding of `vvNew`? If so, why remove this comment?

**Review by dhruv — COMMENTED** [review]
ACK 4676a4fb5be0f6ef0b3f71c1f4361c20f7cb0e0b

A couple nits and questions below.

File: src/addrman.h [hunk 11]
```diff
@@ -525,19 +527,26 @@ friend class CAddrManTest;
...
+
+            // The entry shouldn't appear in more than
+            // ADDRMAN_NEW_BUCKETS_PER_ADDRESS. If it has already, just skip
+            // this bucket_entry.
+            if (info.nRefCount >= ADDRMAN_NEW_BUCKETS_PER_ADDRESS) continue;
+
             int nUBucketPos = info.GetBucketPosition(nKey, true, bucket);
-            if (format >= Format::V2_ASMAP && nUBuckets == ADDRMAN_NEW_BUCKET_COUNT && vvNew[bucket][nUBucketPos] == -1 &&
-                info.nRefCount < ADDRMAN_NEW_BUCKETS_PER_ADDRESS && serialized_asmap_checksum == supplied_asmap_checksum) {
+            if (restore_bucketing && vvNew[bucket][nUBucketPos] == -1) {
```
Comment (ryanofsky) [inline src/addrman.h]: In commit "[addrman] Don't rebucket new table entries unnecessarily" (a5c9b04959f443372400f9a736c6eaf5502284a1)

Just some notes to make sure I'm following this correctly.

- The `restore_bucketing` change here is aesthetic and the only actual change this line is dropping the `format >= Format::V2_ASMAP` condition. 
- The effect of dropping `format` condition is this will now keep existing bucket positions instead of rebucketing when an old format is loaded and there is no asmap. ("Only rebucket if" in commit description means "Use existing bucket positions if")
- In the case where an old format is loaded an there is an asmap, this logic isn't triggered because the `serialized_asmap_checksum == supplied_asmap_checksum` condition is false.
- Maybe could be a unit test checking old format number + no asmap uses  existing bucket positions

**Review by ryanofsky — APPROVED** [review]
Code review ACK 4676a4fb5be0f6ef0b3f71c1f4361c20f7cb0e0b. I'm not previously familiar with this code but all the changes here do make sense and seem like improvements. Left some notes and comments, but they aren't important so feel to ignore.

File: src/addrman.h [hunk 12]
```diff
@@ -523,9 +525,10 @@ friend class CAddrManTest;
             s >> serialized_asmap_version;
         }
 
-        for (int n = 0; n < nNew; n++) {
-            CAddrInfo &info = mapInfo[n];
-            int bucket = entryToBucket[n];
+        for (auto bucket_entry : bucket_entries) {
```
[the remark this answers is not in the document]
Comment (vasild) [reply src/addrman.h:535]: I think `bucket_entries.size() >= nNew`, let me use an example:

```
mapInfo[id=3] = 1.1.1.1
// for the example, I pick up some arbitrary indexes in vvNew
vvNew[2][5] = id=3
vvNew[4][9] = id=3
nNew = 1
```

This would be serialized on disk as:

```
1 (nNew)
0 (nTried)
1024 (new_bucket_count)
1.1.1.1 (list of new addresses) (*)
(zero tried addresses)
0 (number of elements in vvNew[0])
0 (number of elements in vvNew[1])
1 (number of elements in vvNew[2])
0 (the index in (*) of the first non empty element in vvNew[2])
0 (number of elements in vvNew[3])
1 (number of elements in vvNew[4])
0 (the index in (*) of the first non empty element in vvNew[4])
0 (number of elements in vvNew[5])
...
0 (number of elements in vvNew[1023])
```

During unserialization, in the new code from this PR, this will result in:

```
bucket_entries[0] = (bucket = 2, index in (*) = 0)
bucket_entries[1] = (bucket = 4, index in (*) = 0)
```

So, `bucket_entries.size() = 2` while `nNew = 1`.

File: src/addrman.h [hunk 13]
```diff
@@ -504,14 +504,14 @@ friend class CAddrManTest;
...
-                s >> nIndex;
-                if (nIndex >= 0 && nIndex < nNew) {
-                    bucket_entries.emplace_back(bucket, nIndex);
+        for (int bucket = 0; bucket < nUBuckets; ++bucket) {
+            int num_entries{0};
+            s >> num_entries;
+            for (int n = 0; n < num_entries; ++n) {
+                int entry_index{0};
+                s >> entry_index;
+                if (entry_index >= 0 && entry_index < nNew) {
```
[the remark this answers is not in the document]
Comment (vasild) [reply src/addrman.h:511]: That variable `entry_index` describes the index of the address in the disk-only list of new addresses (marked as `(*)` in [the comment above](https://github.com/bitcoin/bitcoin/pull/20557#discussion_r572736572)).

I think that should not be confused with the in-memory-only "id" and `entry_index` describes it well.

File: src/addrman.h (hunk 10 above)
Comment (vasild) [reply src/addrman.h]: No, because the memory-only `id` may not equal to the disk-only `index in the serialized "all new addresses"`.

File: src/addrman.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/addrman.h:535]: This was part of the bug that this PR fixes. nNew is the total number of new addrs. Each new addr can appear in more than one new bucket, so the number of entries in bucket_entries could be higher than nNew. The old code was potentially throwing away these additional placements.

File: src/addrman.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/addrman.h:511]: I think that's a reasonable suggestion, but not worth invalidating the existing ACKs.

File: src/addrman.h [hunk 14]
```diff
@@ -413,13 +414,13 @@ friend class CAddrManTest;
                 }
             }
         }
-        // Store asmap version after bucket entries so that it
+        // Store asmap checksum after bucket entries so that it
         // can be ignored by older clients for backward compatibility.
-        uint256 asmap_version;
+        uint256 asmap_checksum;
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/addrman.h:416]: > (nit) could this is a scripted-diff?

Scripted diffs are useful when there are many occurrences of a string across multiple files. This string is used in 7 places, so is easy enough to review by eye.

File: src/addrman.h (hunk 10 above)
Comment (jnewbery) [reply src/addrman.h]: See the line above. vvNew isn't encoded explicitly but is reconstructed.

File: src/addrman.h [hunk 15]
```diff
@@ -500,7 +500,9 @@ friend class CAddrManTest;
         nTried -= nLost;
 
         // Store positions in the new table buckets to apply later (if possible).
-        std::map<int, int> entryToBucket; // Represents which entry belonged to which bucket when serializing
+        // An entry may appear in up to ADDRMAN_NEW_BUCKETS_PER_ADDRESS buckets,
+        // so we store all bucket-entry_index pairs to iterate through later.
+        std::vector<std::pair<int, int>> bucket_entries;
```
[the remark this answers is not in the document]
Comment (vasild) [reply src/addrman.h:503]: I think that vector could be avoided only if we could seek into the stream. Then we could change the current:
```
read new buckets
read asmap checksum
process the new buckets that were read, depending on the asmap checksum
```
to
```
seek forward and read the asmap checksum (last 32 bytes)
seek back and read the new bucketing data, depending on the asmap checksum
```

[the remark this answers is not in the document]
Comment (jnewbery) [reply src/addrman.h:503]: We only know how we're going to use the vector after we've deserialized the asmap_checksum field, so I don't think we can get rid of it.

File: src/addrman.h [hunk 16]
```diff
@@ -509,7 +511,7 @@ friend class CAddrManTest;
                 int nIndex = 0;
                 s >> nIndex;
                 if (nIndex >= 0 && nIndex < nNew) {
-                    entryToBucket[nIndex] = bucket;
+                    bucket_entries.emplace_back(bucket, nIndex);
```
[the remark this answers is not in the document]
Comment (vasild) [reply src/addrman.h]: A test is in the baking at https://github.com/jnewbery/bitcoin/pull/18.

[the remark this answers is not in the document]
Comment (jnewbery) [reply src/addrman.h]: I definitely agree. @vasild has a test here: https://github.com/jnewbery/bitcoin/pull/18 which can be added after this PR.

File: src/addrman.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/addrman.h:543]: I'll expand this comment in a follow-up.

File: src/addrman.h (hunk 11 above)
Comment (jnewbery) [reply src/addrman.h]: > The restore_bucketing change here is aesthetic and the only actual change this line is dropping the format >= Format::V2_ASMAP condition.

Right, this condition doesn't change during loop iterations, so it's clearer to move it out of the loop, mark it const, and only do a conditional check inside the loop on what can change over iterations.

> The effect of dropping format condition is this will now keep existing bucket positions instead of rebucketing when an old format is loaded and there is no asmap. ("Only rebucket if" in commit description means "Use existing bucket positions if")

That's right. If the file format is old (no asmap), and we don't have an asmap, then we gain nothing by dropping new bucket positions.

> In the case where an old format is loaded an there is an asmap, this logic isn't triggered because the serialized_asmap_checksum == supplied_asmap_checksum condition is false.

Right.

> Maybe could be a unit test checking old format number + no asmap uses existing bucket positions

Perhaps. It's a bit difficult to test because we can't easily serialize addrman in the old format.

**Comment by jnewbery** [comment]
Thanks @dhruv and @ryanofsky for the reviews. I think this is ready for merge now, and the remaining review comments can be done in follow-ups:

- [ ] Add test (https://github.com/bitcoin/bitcoin/pull/20557#issuecomment-737890278, also see https://github.com/jnewbery/bitcoin/pull/18)
- [ ] Use structured bindings to iterate bucket_entries (https://github.com/bitcoin/bitcoin/pull/20557#discussion_r572398580)
- [ ] Expand comment about info.nRefCount >= ADDRMAN_NEW_BUCKETS_PER_ADDRESS (https://github.com/bitcoin/bitcoin/pull/20557#discussion_r572509364)

File: src/addrman.h [hunk 17]
```diff
@@ -523,9 +525,10 @@ friend class CAddrManTest;
             s >> serialized_asmap_version;
         }
 
-        for (int n = 0; n < nNew; n++) {
-            CAddrInfo &info = mapInfo[n];
-            int bucket = entryToBucket[n];
+        for (auto bucket_entry : bucket_entries) {
+            int bucket{bucket_entry.first};
+            const int n{bucket_entry.second};
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/addrman.h]: :heart: structured bindings. Good suggestion. I'll do in a follow-up.

**Comment by laanwj** [comment]
Code review ACK 4676a4fb5be0f6ef0b3f71c1f4361c20f7cb0e0b
